File: it-agents/skills/code-architect/references/code-examples/cpp/Prototype.cpp
// Prototype — copy existing objects via a clone interface without coupling to classes.
#include <iostream>
#include <memory>
#include <string>

class Shape {
protected:
    std::string label;
    int weight = 0;
public:
    virtual ~Shape() = default;
    Shape() = default;
    Shape(std::string l, int w) : label(std::move(l)), weight(w) {}
    virtual std::unique_ptr<Shape> clone() const = 0;
    virtual void describe() const = 0;
};

class RoundShape : public Shape {
    int radius = 0;
public:
    RoundShape(std::string l, int w, int r) : Shape(std::move(l), w), radius(r) {}
    std::unique_ptr<Shape> clone() const override {
        return std::make_unique<RoundShape>(*this); // copy ctor copies all fields.
    }
    void describe() const override {
        std::cout << "RoundShape{" << label << ", w=" << weight << ", r=" << radius << "}\n";
    }
};

class BoxShape : public Shape {
    int side = 0;
public:
    BoxShape(std::string l, int w, int s) : Shape(std::move(l), w), side(s) {}
    std::unique_ptr<Shape> clone() const override {
        return std::make_unique<BoxShape>(*this);
    }
    void describe() const override {
        std::cout << "BoxShape{" << label << ", w=" << weight << ", side=" << side << "}\n";
    }
};

int main() {
    RoundShape original{"prototype-round", 5, 12};
    auto copy = original.clone();
    original.describe();
    copy->describe();

    BoxShape box{"prototype-box", 9, 4};
    box.clone()->describe();
    return 0;
}

File: it-agents/skills/code-architect/references/code-examples/rust/prototype.rs
// Prototype — produce new objects by cloning existing instances rather than
// constructing them from scratch, via a clone-into-box trait.

trait Shape {
    fn clone_box(&self) -> Box<dyn Shape>;
    fn render(&self) -> String;
    fn scale(&mut self, factor: f64);
}

#[derive(Clone)]
struct Circle {
    radius: f64,
}

#[derive(Clone)]
struct Square {
    side: f64,
}

impl Shape for Circle {
    fn clone_box(&self) -> Box<dyn Shape> {
        Box::new(self.clone())
    }
    fn render(&self) -> String {
        format!("Circle(r={:.1})", self.radius)
    }
    fn scale(&mut self, factor: f64) {
        self.radius *= factor;
    }
}

impl Shape for Square {
    fn clone_box(&self) -> Box<dyn Shape> {
        Box::new(self.clone())
    }
    fn render(&self) -> String {
        format!("Square(s={:.1})", self.side)
    }
    fn scale(&mut self, factor: f64) {
        self.side *= factor;
    }
}

fn main() {
    let prototypes: Vec<Box<dyn Shape>> =
        vec![Box::new(Circle { radius: 2.0 }), Box::new(Square { side: 3.0 })];

    for proto in &prototypes {
        let mut copy = proto.clone_box();
        copy.scale(2.0);
        println!("original {} -> cloned&scaled {}", proto.render(), copy.render());
    }
}
